// lemonade/kernel/links/httpd.js
/******************************************************************************
 * Lemonade JS
 * https://lemonadejs.com
 *
 * ----
 *
 * Http daemon link loader script
 *
 *****************************************************************************/

function httpdLoader () {}

/**
 * Adaptor reference
 */
httpdLoader.prototype.adaptor;

/**
 * Create the object and start the server
 * @param {number} port
 * @param {function} cb
 */
httpdLoader.prototype.load = function(port, cb) {
    this.adaptor = this.kernel.include(this.kernel.kerneldir + 'daemons/httpd').singleton('httpd', port);
    this.kernel.on(this.adaptor.ONLINE, function() {
        cb(null, 'listening on ' + port);
    });
};

/**
 * Reference to the httpd server object
 * @return {}
 */
httpdLoader.prototype.link = function() {
    return this.adaptor;
};

/**
 * Status
 * @return {}
 */
httpdLoader.prototype.status = function() {
    return this.adaptor.status;
};

/**
 * Unload
 * @param {function} cb
*/
httpdLoader.prototype.unload = function(cb) {
    if (!this.adaptor) {
        cb(null, 'not listening');
    } else {
        this.adaptor.shutdown(function(err) {
            cb(err, 'stopped listening');
        });
    }
};

/**
 * Exports
 */
exports = module.exports = httpdLoader;
