// lemonade/kernel/links/rabbit.js
/******************************************************************************
 * Lemonade JS
 * https://lemonadejs.com
 *
 * ----
 *
 * Rabbitmq link loader script
 *
 *****************************************************************************/

function rabbitLoader() {}

/**
 * Pool reference
 * @type {object}
 */
rabbitLoader.prototype.pool;

/**
 * Load
 * @param {object} config
 * @param {function} cb
 */
rabbitLoader.prototype.load = function(config, cb) {
    if (!config) {
        throw new Error('Rabbit invalid config');
    }
    /**
     * Create a singleton of the pool
     */
    this.pool = this.kernel.include(this.kernel.kerneldir + 'adaptors/rabbit')
               .singleton('rabbit', config);
    /**
     * Fetch a connection from the pool to test it out
     */
    this.pool.connection(function(err, conn) {
        conn.release();
        cb(err, '- rabbitmq pool up');
    }.bind(this));
};

/**
 * Kernel link
 * @param {function} cb
 */
rabbitLoader.prototype.link = function(cb) {
    this.pool.connection(cb);
};

/**
 * Status object
 */
rabbitLoader.prototype.status = function() {
    return this.pool.status();
};

/**
 * Shutdown
 * @param {function} cb
 */
rabbitLoader.prototype.unload = function(cb) {
    this.pool.shutdown(function() {
        cb(null, '- rabbitmq pool drained');
    });
};

/**
 * Export
 */
exports = module.exports = rabbitLoader;
