// lemonade/kernel/links/spinner.js
/******************************************************************************
 * Lemonade JS
 * https://lemonadejs.com
 *
 * ----
 *
 * Spinner daemon link loader script
 *
 *****************************************************************************/

function spinnerLoader () {}

/**
 * Adaptor reference
 * @type {object}
 */
spinnerLoader.prototype.adaptor;

/**
 * Create the object and start spinning
 * @param {function} cb
 */
spinnerLoader.prototype.load = function(cb) {
    this.adaptor = this.kernel.include(this.kernel.kerneldir + 'daemons/spinner').singleton('spinner',
        function(err) {
            cb(err, 'spinning')
        }
    );
};

/**
 * Reference to the spinner object
 * @return {object}
 */
spinnerLoader.prototype.link = function() {
    return this.adaptor;
}

/**
 * Status
 * @return {object}
 */
spinnerLoader.prototype.status = function() {
    return this.adaptor.status();
};

/**
 * Unload
 * @param {function} cb
 */
spinnerLoader.prototype.unload = function(cb) {
    if (!this.adaptor) {
        cb(null, 'never spun');
    } else {
        this.adaptor.shutdown(function(err) {
            cb(err, 'stopped spinning');
        });
    }
};

/**
 * Exports
 */
exports = module.exports = spinnerLoader;
